### ISSUE TITLE
Allow to run bytecode instrumentation after all compilation tasks

### ISSUE THREAD

**Issue by nskvortsov** [gradle/gradle#2137]
### Expected Behavior
Allow to transform `.class` files, created by (potentially multiple) compilation tasks as a part of building the sourceSet

### Current Behavior
At the moment, all `class` files are generated directly into `sourceSet.output.classesDir`. Instrumenting them in place breaks Gradle incremental builds.
Instrumenting as a last operation of compilation task is not an option in case of Kotlin-Java mixed projects, as CompileKotlin task only generates `.class` for Kotlin sources, leaving Java sources to javac, while instrumentation needs all classes together.

### Context
When attempting to migrate a project from IntelliJ IDEA to Gradle (as a build tool), one needs to apply Jetbrains code instrumentation to have nullability checks and UI forms (for IntelliJ plugins) inserted into bytecode. The instrumenter is available as Ant task.

**Comment by oehme** [comment]
Byte code instrumentation is a widespread pattern in the JVM ecosystem, it should be a first-class citizen supported by the java-base plugin.

**Comment by nskvortsov** [comment]
Gradle 4.0 introduced multiple destination directories for sourceSets. Working around this to apply bytecode postprocessing is getting even trickier with this change.

**Comment by nskvortsov** [comment]
Another project that could really use this feature: https://github.com/Kotlin/kotlinx.atomicfu

**Comment by nskvortsov** [comment]
One more project, that could use instrumentation as a part of build process:

https://gitlab.ow2.org/asm/asm

Their current workaround is doing instrumentation as a "doLast" of compile task. Real problem here is that instrumenting code itself is a part of the build, so some sources have to be compiled three times per build.

**Comment by nskvortsov** [comment]
And another project, that really could use some help: https://github.com/jetbrains/kotlin

Extensive usage of `org.jetbrains.annotations.NotNull` and `Nullalbe` requires resulting bytecode instrumentation (to add runtime checks). Current workaround includes direct manipulation `sourceSet.output.classesDirs` and `sourceSet.builtBy` and some other parts [here](https://github.com/JetBrains/kotlin/blob/5c8afea68ea06f2787f8b96c2b1a9a7631741c33/buildSrc/src/main/kotlin/instrument.kt#L61)

**Comment by oehme** [comment]
> Instrumenting as a last operation of compilation task is not an option in case of Kotlin-Java mixed projects, as CompileKotlin task only generates .class for Kotlin sources, leaving Java sources to javac

This is a performance issues on its own and should be fixed. The Kotlin compiler should compile the Java code itself.

> instrumentation needs all classes together.

Why does it? The Kotlin-compiled classes are on the classpath of the JavaCompile task, so it can look up information about these types. Handling the classpath is already necessary if an instrumenter wants to handle cross-project and external dependencies correctly. So this should just work. What am I missing?

**Comment by oehme** [comment]
If you do nullable analysis, you need to consider the classes on your classpath already. Otherwise how would you know whether a call to a library is safe? So your instrumenter must already work with the classpath. 

Making instrumentation a separate task would mean copying all classes, wasting space and IO time. They should be modified in place as part of the compile task instead. `doLast` is perfectly viable for that. We may add convenience in the future, but it will still be part of the compile tasks's execution.

I don't see any instrumentation in the asm case you linked. They just include the same sources into three projects.

**Comment by nskvortsov** [comment]
> If you do nullable analysis, you need to consider the classes on your classpath already
Indeed, but you also need to understand, which class files are to be transformed. In Kotlin-Java case as it is, part of these classes is missing in compile output directory.

The actual instrumentation code is located [here](https://gitlab.ow2.org/asm/asm/blob/master/build.gradle#L215)

**Comment by oehme** [comment]
> Indeed, but you also need to understand, which class files are to be transformed. In Kotlin-Java case as it is, part of these classes is missing in compile output directory.

For now this can still be worked around I think - You can instrument the Kotlin-generated classes in a `doLast` on the KotlinCompile task, using Java stubs on the classpath so you have the nullable signatures for the Java code. For the Java-generated classes you can do a `doLast` on the JavaCompile task, using the Kotlin-generated classes on the classpath.

All of this would be simpler if the KotlinCompile task did all its work in one atomic unit instead of spreading things out over 2-4 tasks. It would also be way faster, but that's another discussion entirely.

**Comment by kb-1000** [comment]
As this issue is already 3 years old... will you ever implement it?

**Comment by mauromol** [comment]
Still desirable.

**Comment by Col-E** [comment]
If #9041 is to go through, what exactly would be a _"nice"_ way for such instrumenting plugin to handle itself?

- Overwriting the `classes` dir contents was easy and _"works"_ but prevents builds from being incremental, and may be blocked in the future
- Writing instrumented classes to a new dir `instClasses` works, but then existing tasks that use `classes` as an input will not see the instrumented classes by default
    - Hacking `jar` to only pull in the instrumented classes does not seem feasible from my experimentation, but even if it is seems like a hack. If somebody knows how to get this to work I'd love to hear it though.
    - Adapting complex builds to not use the implicit `jar` task adds a lot of required boilerplate for something that should ideally be a _"set and forget"_

**Comment by iseki0** [comment]
@oehme 
> Making instrumentation a separate task would mean copying all classes

But what will happends if we don't copy the classes and just in-place modify the class file? I means use the same directory(the compilation output) as the instrumentation task's input and output.

> An instrumenter instrumenting itself is a very special case

I don't think so, since graalvm native-image is lack of dynamic capabilities. We need more instrumentation step to fill the requirements of it. But I'm agreed with need not to change the build pipeline. Gradle is a widely used build tools.

**Comment by iseki0** [comment]
@oehme 
> An instrumenter instrumenting itself is a very special case

I don't think so, since the Graalvm native-image is lack of dynamic capabilities. We need more and more post-processing stage to fill its requirements.

**Comment by big-guy** [comment]
This is something we're still interested in doing somehow, but the current compilation model (compile tasks, source sets and source set outputs) make this hard to wire in as-is. 

We'd prefer if there was a higher level model that allowed plugins to contribute to the compilation process and insert themselves at the right time.

At the moment, there are no concrete plans to address this particular use case, but we are looking at how to pull up some use cases to a higher level model.

**Comment by iseki0** [comment]
@big-guy What is the "high level model"? May I interpret that you're starting finding a solution?
